hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/BatchStage.java

```diff
@@ -468,6 +468,51 @@
     @Nonnull @Override
     <R> BatchStage<R> customTransform(@Nonnull String stageName, @Nonnull ProcessorMetaSupplier procSupplier);
 
+    /**
+     * Transforms {@code this} stage using the provided {@code transformFn} and
+     * returns the transformed stage. It allows you to extract common pipeline
+     * transformations to a separate function and then chain usage of those
+     * functions.
+     * <p>
+     * For example say you have this pipeline:
+     *
+     * <pre>{@code
+     *     p.drawFrom(...)
+     *       .map(... some mapping)
+     *       .flatMap(... some mapping)
+     *       ...
+     * }</pre>
+     *
+     * You can extract the {@code map} and {@code flatMap} stages to a
+     * function:
+     *
+     * <pre>{@code
+     *     BatchStage<String> addSpecialMapping(BatchStage<String> inputStage) {
+     *          return inputStage
+     *              .map(... some mapping)
+     *              .flatMap(... some mapping);
+     *     }
+     * }</pre>
+     *
+     * And then use it in the following way:
+     *
+     * <pre>{@code
+     *     p.drawFrom(...)
+     *       .apply(this::addSpecialMapping)
+     *       ...
+     * }</pre>
+     *
+     * The {@code addSpecialMapping} method can then be re-used in multiple
+     * pipelines.
+     *
+     * @param transformFn a function to transform this stage to another stage
+     * @param <R> type of the returned stage
+     */
+    @Nonnull
+    default <R extends Stage> R apply(@Nonnull FunctionEx<BatchStage<T>, ? extends R> transformFn) {
```

Comment:
It should be `? super T`

Reply:
I ended up adding `? super BatchStage`. This way the actual parameter can be for example `FunctionEx<BatchStage<? extends Number>, Whatever>`. Without it, the function input has to be `BatchStage<Integer>`.

I will not pretend that I understand why, @mtopolnik might be able to explain it :).